### PR TITLE
`api_key` param cannot override environmental GROQ_API_KEY for groq

### PR DESCRIPTION
**Pull request by Raymond1415926**
## Title

`api_key` param cannot override environmental GROQ_API_KEY for groq

## Relevant issues

None

## Type

🐛 Bug Fix

## Changes

In main.py:
image

## [REQUIRED] Testing - Attach a screenshot of any new tests passing locall
In **itellm/tests/test_api_key_override_dynamic_key.py**
==========invalid key test============
Give Feedback / Get Help: https://github.com/BerriAI/litellm/issues/new
LiteLLM.Info: If you need to debug this error, use `litellm.set_verbose=True'.

Provider List: https://docs.litellm.ai/docs/providers

litellm.BadRequestError: GroqException - Error code: 401 - {'error': {'message': 'Invalid API Key', 'type': 'invalid_request_error', 'code': 'invalid_api_key'}}
==============try original===============
ModelResponse(id='chatcmpl-c661215d-1649-41d3-9627-b0c997d9b93e', choices=[Choices(finish_reason='stop', index=0, message=Message(content="Hello! I'm just an AI language model, so I don't have feelings or personal experiences, but I'm here and ready to assist you with any questions or tasks you have. How can I help you today?\n\nIf you're interested, you can ask me about a wide range of topics such as science, technology, literature, history, and more. I can also help you with tasks like generating ideas, drafting emails, or just having a friendly conversation. What would you like to do?", role='assistant', tool_calls=None, function_call=None))], created=1721096646, model='groq/mixtral-8x7b-32768', object='chat.completion', system_fingerprint='fp_c5f20b5bb1', usage=Usage(completion_tokens=108, prompt_tokens=18, total_tokens=126))

1 test case for no having invalid key as override, triggering invalid_api_key error
1 test case for no key overrride, successfully use environmental key

### CODE REVIEW

File: litellm/main.py [hunk 1]
```diff
@@ -910,7 +910,7 @@ def completion(
             llm_provider=custom_llm_provider, dynamic_api_key=api_key
         )  # get the api key from the environment if required for the model
 
-        if dynamic_api_key is not None:
+        if dynamic_api_key is not None and api_key is None:
```
Comment (ghost) [inline litellm/main.py]: @Raymond1415926 this is the wrong place for the fix. 

It should be in get_llm_provider - https://github.com/BerriAI/litellm/blob/9e07af6c0245ba19696aea27b6a94ee1f587fe97/litellm/utils.py#L4274

**Review by ghost — CHANGES_REQUESTED** [review]
Please make requested changes

File: litellm/main.py (hunk 1 above)
Comment (Raymond1415926) [reply litellm/main.py]: Can I make it global? Or it has to be case by case?

Other providers, by providing the same get_secret logic, will result in the same issue where input api_key never overrides the environmental variables.

Comment (ghost) [reply litellm/main.py]: i think a global approach is fine

File: litellm/utils.py [hunk 2]
```diff
@@ -4249,6 +4249,7 @@ def get_llm_provider(
             model.split("/", 1)[0] in litellm.provider_list
             and model.split("/", 1)[0] not in litellm.model_list
             and len(model.split("/"))
+            and not api_key
```
Comment (ghost) [inline litellm/utils.py]: hey @Raymond1415926 this isn't what i thought you meant, when you said 'applied globally'

please add the check for each provider in the block

Comment (ghost) [reply litellm/utils.py]: reference implementation: https://github.com/BerriAI/litellm/blob/eaee90b02e6a0a9a20ddad5d9bf70136799511db/litellm/utils.py#L4304

**Comment by Raymond1415926** [comment]
@krrishdholakia Hey, just changed.

Also attached additional test cases for all the providers that I added key override

**Comment by Raymond1415926** [comment]
@krrishdholakia Hey, just following up on this. Did you see my latest update?

**Comment by Raymond1415926** [comment]
Updates:

I fixed a bug when tool_calls is None, token_counter will throw error saying NoneType is not iterable.

Tests:

test_empy_tool_calls.py

**Pull request closed without merging**